### PR TITLE
MLX: DoRA training support in get_peft_model

### PR DESCRIPTION
**Pull request by Lyxot**
### What this fixes

Requesting DoRA on Apple Silicon silently trained plain LoRA. The MLX `get_peft_model` had no `use_dora` parameter, so the flag was absorbed by its `**kwargs` catch-all and dropped — no error, no warning, and an adapter that looks like a normal LoRA. Save-time magnitude collection, `fine_tune_type="dora"` stamping and DoRA reload already existed; the only missing piece was wrap time, which is what this adds.

### What changed

`use_dora` is now a named parameter on the MLX `get_peft_model`. When set, targets are wrapped with mlx-lm's `DoRALinear` and each wrapper's magnitude vector is unfrozen so it actually trains — without that, the run learns direction only and is DoRA in shape but not in what it optimises. The parameter is appended after the existing ones, so positional callers keep binding what they always did, and it gets the same strict `bool` check `use_rslora` already has.

Three constraints are worth calling out for review:

**Exact base types only.** DoRA wraps `nn.Linear` and `nn.QuantizedLinear` and refuses anything else by name. Subclasses are not accepted even though `isinstance` would allow them: a fused projection returns a tuple of splits that its caller destructures, while `DoRALinear` returns a single array, so wrapping one attaches cleanly and then breaks the next forward. For the same reason the module's own `to_lora()` shortcut is skipped under DoRA, since it produces plain-LoRA wrappers for exactly those fused layers.

**Refusal is all-or-nothing.** Every selected target across the layer walk and the root pass is type-checked before any replacement is committed, so an unsupported base cannot leave earlier layers converted and the suggested retry with `use_dora=False` still works on the tree that comes back. A wrapper construction that fails afterwards for some other reason is not covered by that guarantee.

**Two requests are refused up front.** A vision or projector tower, because those wrap as plain LoRA and an adapter records one `fine_tune_type` for the whole tree — the mixed result saves without complaint, is stamped as DoRA, and is rebuilt entirely as DoRA on reload, bringing the plain-LoRA layers back as a wrapper they were never trained as. And `init_lora_weights=False`, because the magnitude is seeded from the base weight norm, which describes the adapted weight only while `lora_b` is zero.

Nonzero `lora_dropout` is deliberately **not** refused. It trains normally in MLX format; what differs is that peft applies DoRA dropout inside the magnitude correction while mlx-lm leaves the base undropped, so continued training under peft would not reproduce the same run. Wrap time prints that caveat and proceeds, since an interchange limitation should not block a native run.

### Validation

Coverage joins the existing routed-expert wrapper-selection tests, next to the fused-linear dispatch case the exact-type rule exists for: exact-type wrapping with the magnitude receiving a gradient, quantized-base construction, atomic refusal across both the layer walk and the root pass, the request refusals, and plain-LoRA positional compatibility. Two of those were mutation-checked — reverting the unfreeze gate to a plain module walk, and removing the root pre-validation, each fail their test.

Beyond the suite, on a real model: wrap, train, save and reload with bitwise-equal magnitudes and `fine_tune_type="dora"`; checkpoint resume carrying the new magnitude optimiser state; a quantized (QLoRA) base; and plain LoRA unchanged.

End to end through Studio, a 15-step run on `unsloth/Qwen3.5-0.8B` produced an adapter with `fine_tune_type: "dora"` and one magnitude per adapted module (114/114), and every magnitude had moved off its `‖W‖` seed — so both parts of the decomposition are learning, not just the direction.

Full suite shows no new failures against the same base.

### Notes for reviewers

- Magnitude unfreezing is restricted to modules owning a complete LoRA pair and then to the DoRA wrapper classes, the same two-part gate `collect_mlx_lora_adapter_tensors` already uses, so a base module that merely owns an `m` parameter is never unfrozen. It also runs only for DoRA requests, leaving the plain-LoRA path untouched.
- DoRA on a routed/MoE base is refused, since mlx-lm has no DoRA wrapper for `SwitchLinear`.
- Embeddings never reach a wrapper here — embedding targets go to full-module training — so there is nothing to expose or refuse for them.

A companion change in the main repository lets Studio request this. It refuses at run start when the installed library predates this parameter, so the two can land in either order.

### CODE REVIEW

File: unsloth_zoo/mlx/loader.py [hunk 1]
```diff
@@ -7655,10 +7734,57 @@ def get_peft_model(
...
+                # undropped, so the same adapter learns differently under
+                # the two. Say so now rather than blocking an MLX-only run.
+                print(
+                    "Unsloth: DoRA with lora_dropout > 0 trains normally in "
+                    "MLX format, but peft applies DoRA dropout inside the "
+                    "magnitude correction while mlx-lm leaves the base "
+                    "undropped, so continuing this adapter under peft would "
+                    "not reproduce the same training. Use lora_dropout=0 to "
+                    "keep the two equivalent."
+                )
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/loader.py:8082]: Keeping `print()` here, for consistency with the surrounding notices.

Inside `get_peft_model` there are 4 `print("Unsloth: ...")` notices against 1 `warnings.warn`, and the closest precedent is the same class of message: `Unsloth: bias is not supported for MLX LoRA yet - ignoring bias=...` — an option accepted, with a caveat reported. The module does use `warnings.warn` in 11 places, but for stronger conditions: a dtype that will not be honored, a forced requantization, or a configuration that will apply no LoRA at all. This one is the weaker kind — training proceeds exactly as asked, and only interchange with peft is affected.

Switching just this line would make it the only notice of its kind in the function using a different mechanism. Moving all of them to `warnings.warn` is a reasonable change and I would support it, but it is a module-wide convention decision rather than something to fold into this PR.

File: unsloth_zoo/mlx/loader.py [hunk 2]
```diff
@@ -307,6 +366,20 @@ def linear_to_lora_layers(model, num_layers, config):
...
+        # failing on those existing DoRA wrappers. Checking every selected
+        # target first makes the TYPE refusal all-or-nothing; a construction
+        # that fails afterwards for another reason is not covered.
+        for layer in layers[max(len(layers) - limit, 0):]:
+            for name, module in layer.named_modules():
+                if name in keys:
+                    _mlx_dora_wrapper_type(module, name)
+        for name, module in model.named_modules():
+            if name in keys:
+                _mlx_dora_wrapper_type(module, name)
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/loader.py:366]: Measured this before deciding. On `unsloth/Qwen3.5-0.8B` (489 modules, 24 layers) the pre-validation pass validates 162 targets in **2.10 ms** (best of 5) against a 4.1 s model load — 0.05% of load time. The pass touches module objects only; no tensor work happens in it.

```text
targets validated:    162
pre-validation pass:  2.10 ms (best of 5)
model load:           4.1 s
share of load time:   0.0511%
```

Collecting `(owner, name, module)` once would mean restructuring the traversal in `linear_to_lora_layers`, which is the shared path the plain-LoRA case also uses and which this PR deliberately leaves unchanged. For a 0.05% cost that is not a trade I would make here.

File: unsloth_zoo/mlx/loader.py [hunk 3]
```diff
@@ -271,7 +271,66 @@ def _mlx_lora_spec_for_module(module, specs):
...
+    module that merely owns an `m` parameter is never unfrozen. Without this
+    the magnitudes stay frozen and only the direction is learned.
+    """
+    from .utils import iter_mlx_lora_modules
+
+    count = 0
+    for _name, module in iter_mlx_lora_modules(model):
+        if hasattr(module, "m") and type(module).__name__.startswith("DoRA"):
+            module.unfreeze(keys=["m"], recurse=False)
+            count += 1
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/loader.py:283]: This gate matches four existing sites in the MLX backend that use exactly `type(module).__name__.startswith("DoRA")` — including `collect_mlx_lora_adapter_tensors` in `mlx/utils.py`, which decides which tensors get written into the adapter. An `isinstance(module, DoRALinear)` check here would be inconsistent with those, and would also miss `DoRAEmbedding`, which mlx-lm ships.

The false positive you describe is already excluded by the other half of the gate: the walk iterates `iter_mlx_lora_modules(model)`, so a module has to own a complete `lora_a`/`lora_b` pair before its class name is considered at all. There is a test for exactly that scenario — a `DoRAttention` class owning an `m` parameter but no adapter — asserting it stays frozen.

A rename upstream would break all four sites together, so it wants one coordinated change rather than a divergent check in the newest one.

File: tests/test_mlx_switch_lora.py [hunk 4]
```diff
@@ -555,3 +556,168 @@ def test_legacy_switch_rank_layout(tmp_path):
...
+
+def test_dora_refuses_fused_and_routed_bases_atomically():
+    """The to_lora() shortcut yields plain-LoRA wrappers for fused linears
+    (see test_custom_to_lora_dispatch_preserves_fused_outputs), so DoRA
+    skips it and wraps EXACT Linear types only: a fused projection returns
+    a tuple its caller destructures, while DoRALinear returns one array."""
+    import mlx.nn as nn
+    from mlx_lm.models.afm7 import FusedLinear
+    from mlx_lm.models.switch_layers import SwitchLinear
+    from unsloth_zoo.mlx.loader import (
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply tests/test_mlx_switch_lora.py:623]: The same file already hard-imports this at line 257 — `from mlx_lm.models.afm7 import FusedLinear, FusedLoRALinear` — for the sibling fused-dispatch test that the exact-type rule exists for. The new test follows it, and adding `importorskip` to only the newer of the two would leave the pair inconsistent.

It is also not reachable in the supported range: `afm7` landed in mlx-lm v0.26.0, and this package floors at `mlx-lm>=0.28.3`, so the module is present in every version installed. The file additionally skips at module level when the real MLX runtime is not available.

**Comment by BardiaKoopah** [comment]
Reviewed this on an M2 16GB MacBook Pro.

### Reproducing the bug

Confirmed on the base commit. `use_dora=True` returns a `LoRALinear` with trainable keys `{lora_a, lora_b}` — no exception, no warning, nothing in the output to suggest the request was dropped. Silent, exactly as described.

On this branch the same call gives a `DoRALinear` with `{lora_a, lora_b, m}`.

### Verifying the magnitude actually learns

This was the part I most wanted to check independently, since a magnitude that's merely *present* looks identical to one that's training until you measure it.

- `m` is seeded from `‖W‖` — verified against `mx.linalg.norm(W, axis=1)`
- It receives a gradient: `|grad m|` = 54.94 → 9.54 → 1041.55 across three SGD steps
- It **moves**: `max|m_after − m_before|` = 12.156
- The base weight stays frozen throughout

So both halves of the decomposition are being optimised, not just the direction.

### Suite

M1 CI file set from `.github/workflows/mlx-ci.yml`, one pytest invocation per file, run against both the merge base and this branch:

| | base `9aa506ea` | this branch |
|---|---|---|
| **Total** | 124 passed, 0 failed | **127 passed, 0 failed** |

The +3 is exactly your new DoRA tests. No regressions, no flakes on either side. Also merges cleanly into current `main`.

### Beyond your tests

A few edges your tests don't reach that I probed separately, all of which came back correct:

- **VLM language tower only** — wraps as `DoRALinear` with `m` trainable, vision tower correctly left at `Linear`. No silent downgrade on the VLM path.
- **`embed_tokens` in `target_modules`** — routed to full-module training as you describe; never reaches the wrapper, so the refusal can't misfire on it.
- **Quantized base** — `DoRALinear(QuantizedLinear)`, `m.shape=(16,)`, trainable.
- **Save → reload round trip** — `m` written to `adapters.safetensors`, config stamped `dora`, and both `m` and `lora_a` come back **bitwise-equal** after reloading into a fresh base model.
- **Gradient checkpointing on** — worth calling out because every DoRA test here passes `use_gradient_checkpointing=False`, while real runs default to `True`. It works: `|grad m|` = 38.70 and `m` still moves. Given what #963 documents about checkpointing silently dropping attribute-carried gradients, I wanted to be sure that failure mode didn't apply — it doesn't, since the magnitude is a real parameter on the wrapper.

Two things I checked by instrumenting rather than reading, because the source alone wasn't proof:

- **The reload path can't fire the new DoRA branch.** A dense DoRA reload makes **zero** calls to `_mlx_lora_from_base` — it routes through `DoRALinear_cls` instead. The only shape that does reach it (a `dora`-stamped config containing a routed `SwitchLinear`) returns `LoRASwitchLinear` without raising.
- **The root pass never receives an already-wrapped module.** Across four model shapes including a deliberately aliased one, all four `_mlx_dora_wrapper_type` calls received a plain `Linear`. The half-converted-tree scenario looks unreachable.

One thing I want to explicitly *not* report: aliased modules split into two independent wrappers. I checked it against the base commit first and it's identical there for plain LoRA, so it's pre-existing and has nothing to do with this change.

### Three questions

**1.** The guard is `type(module) in (nn.Linear, nn.QuantizedLinear)`, so it fires for any other type, but the message mentions only fused projections and routed experts. I couldn't construct a case where that misleads — embeddings are routed away before reaching it — so this is cosmetic. Mentioning it only in case you see a path I don't.

**2.** `_unfreeze_dora_magnitudes` returns a count that both call sites discard. A zero under `use_dora=True` would mean DoRA was requested and no magnitude is training, which is the same silent failure this PR exists to fix. I couldn't reach it given the pre-validation, so this is defensive rather than a bug — but an assertion there would be cheap insurance.

**3.** The one with a concrete action: all three DoRA tests pass `use_gradient_checkpointing=False`. The behaviour is correct at `True` (I checked), but the suite wouldn't catch a regression there, and checkpointing is the default in real runs. A `gc=True` case seems worth adding.

Nice change — the exact-type rule and the all-or-nothing pre-validation both held up under everything I threw at them.

**Comment by Lyxot** [comment]
@codex review

**Comment by Lyxot** [comment]
Thanks for going this deep — reproducing the downgrade on the base commit and measuring the magnitude gradient independently is exactly the part that can't be taken on faith, since a present-but-frozen `m` looks identical to a training one until someone measures it.

Heads up on the force-push: the branch is now rebased onto current main (110 commits). The commit's content is byte-identical to what you reviewed (`git range-diff` reports `=`), so your run still describes this code. I also re-checked it against the two subsystems main has gained since the old base, because a clean rebase says nothing about those: the quantized optimizer state gates on element count rather than shape, so a 1-D `m` takes the full-width path below 4096 elements and the quantized path at 4096 — verified in both regimes, and at lr=1e-3 the quantized and plain optimizers agree to four decimals. The base weight stays bitwise identical while `m` moves, and the seed still matches `‖W‖` exactly.

On your three points:

**1. Refusal message wording** — leaving it. The message names the rule before the examples ("covers plain Linear/QuantizedLinear bases only, so fused projections and routed experts are unsupported"), so the general case is stated even when the specific base is neither of those, and you found no input it actually misleads.

**2. Discarded magnitude count** — not adding the assertion, because the state it would catch isn't reachable. `_raise_no_lora_targets` runs before `_unfreeze_dora_magnitudes` on both the VLM and text paths, so a wrap that attached nothing has already failed the run; and the all-or-nothing type check accepted every selected target before any wrapper was committed, so anything that did attach is a DoRA wrapper. A zero count would mean one of those two guarantees was already broken, and asserting on it there would read as doubting the pre-validation rather than checking something new. The tests already pin the positive direction — one magnitude per adapted module, each trainable and each moving off its seed.

**3. Gradient checkpointing** — you're right that this is the real gap: every DoRA test passes `use_gradient_checkpointing=False` while real runs default to `True`. I've verified the behaviour rather than added a case, driving it through the backend's own `apply_gradient_checkpointing` (the layer-class `mx.checkpoint` patch) instead of a local approximation: the `m` gradient survives (`sum|grad m|` 0.38 → 0.22 → 0.16 over three steps) and the magnitude moves 1.2e-1, matching your reading. Confirming the `#963` failure mode doesn't apply here for the reason you identified — `m` is a real parameter on the wrapper, not an attribute carried past the rewriter. Keeping it out of the suite is deliberate: it's a coverage gap rather than a behaviour change, and this branch is held to a small diff, so the proof lives in recorded validation. Worth a follow-up that covers the checkpointed path for the adapter wrappers generally, not just DoRA.

**Comment by danielhanchen** [comment]
Confirmed the MLX get_peft_model in loader.py still swallows use_dora through its kwargs catch-all, so the request trains plain LoRA with no notice, and this adds the missing wrap-time piece. Will get this reviewed.

File: unsloth_zoo/mlx/loader.py [hunk 5]
```diff
@@ -317,6 +362,18 @@ def linear_to_lora_layers(model, num_layers, config):
...
+        # found late would leave earlier layers converted and the retry
+        # advice (use_dora=False) failing on them. Only the TYPE refusal is
+        # all-or-nothing; a later construction failure is not covered.
+        for layer in layers[max(len(layers) - limit, 0):]:
+            for name, module in layer.named_modules():
+                if name in keys:
+                    _mlx_dora_wrapper_type(module, name)
+        for name, module in model.named_modules():
+            if name in keys:
+                _mlx_dora_wrapper_type(module, name)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/loader.py]: Correct, and reachable: with role-based selection the conversion narrows keys per layer via layer_keys and at the root via shared, while the preflight still walked the union. On a two-layer model whose local proj is a dense Linear in layer 0 and a FusedLinear in layer 1, plain LoRA attaches 1 and leaves layer 1 alone, but use_dora=True refused the whole run. Fixed in c84c0e70 by mirroring both predicates, with a regression test that fails against the previous validation walk. A selected unsupported base is still refused, and still before any wrapper is committed.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
